src/koheesio/spark/transformations/download_files.py

```diff
@@ -0,0 +1,227 @@
+
+    | url                          | downloaded_file_path |
+    |------------------------------|----------------------|
+    | http://example.com/file1.txt | download/file1.txt   |
+    | http://example.com/file2.txt | download/file2.txt   |
+
+    Parameters
+    ----------
+    column : Union[Column, str]
+        The column that holds the URLs to download.
+    download_path : str
+        The local directory path where the file will be downloaded to.
+    chunk_size : int, optional, default=8192
+        The size (in bytes) of the chunks to download the file in, must be greater than 16.
+    mode : FileWriteMode, optional, default=FileWriteMode.OVERWRITE
+        Write mode: overwrite, append, ignore, exclusive, or backup.
+
+    Write Modes
+    -----------
+    The `DownloadFileFromUrlTransformation` supports the following write modes:
+
+    - OVERWRITE
+    - APPEND
+    - IGNORE
+    - EXCLUSIVE
+    - BACKUP
+    """
+
+    column: Union[Column, str] = Field(
+        default="",
+        description="The column that holds the URLs to download.",
+    )
+    download_path: DirectoryPath = Field(
+        ..., description="The local directory path where the file will be downloaded to."
+    )
+    chunk_size: int = Field(
+        8192,
+        ge=16,
+        description="The size (in bytes) of the chunks to download the file in, must be greater than or equal to 16.",
+    )
+    mode: FileWriteMode = Field(
+        default=FileWriteMode.OVERWRITE,
+        description="Write mode: overwrite, append, ignore, exclusive, or backup.",
+    )
+
+    class Output(ColumnsTransformationWithTarget.Output):
+        download_file_paths: dict[str, str] = Field(
+            default_factory=dict, description="The downloaded file paths per URL."
+        )
+
+    def func(self, partition: set[str]) -> None:
+        """
+        Takes a set of urls and downloads the files from the URLs in the specified column.
+
+        Parameters
+        ----------
+        partition : set[str]
+            A set of URLs to download the files from.
+        """
+        for url in partition:
+            step = DownloadFileStep(
+                url=url, download_path=self.download_path, mode=self.mode, chunk_size=self.chunk_size
+            )
+            step.execute()
+            self.output.download_file_paths[url] = step.output.download_file_path.relative_to(
+                self.download_path.parent
+            ).as_posix()
+
+    def execute(self) -> Output:
+        """
+        Download files from URLs in the specified column.
+        """
+        # Collect the URLs from the DataFrame and process them
+        source_column_name = self.column
+        if not isinstance(source_column_name, str):
+            source_column_name = get_column_name(source_column_name)
+       
+        partition = {row.asDict()[source_column_name] for row in self.df.select(self.column).collect()}  # type: ignore
+        self.func(partition)
```

Comment:
Is it maybe an idea to rename this function to something more descriptive? Because my initial thought when i saw this name was: "This is so generic, it must be a bespoke build-in function from Koheesio/Spark."

Reply:
execute is a built in of Koheesio. Since this is extends Transformation (`ColumnsTransformationWithTarget`), it also has a `transform` method.

So the way to interact with this class is like this:

```python
# using transform method
transformation = DownloadFileFromUrlTransformation(
    column="url",
    target_column="downloaded_file_path",
    mode=FileWriteMode.OVERWRITE,
    download_path="/path/to/download"
)
transformed_df = transformation.transform(df)
```

```python
# using execute method
transformation = DownloadFileFromUrlTransformation(
    df=df
    column="url",
    target_column="downloaded_file_path",
    mode=FileWriteMode.OVERWRITE,
    download_path="/path/to/download"
)
transformed_df = transformation.execute().df
```

```python
# using df.transform
download_urls = DownloadFileFromUrlTransformation(
    column="url",
    target_column="downloaded_file_path",
    mode=FileWriteMode.OVERWRITE,
    download_path="/path/to/download"
)
df = df.transform(download_urls)
```

In other words: standard Koheesio way of working :)

Reply:
I might have mis spoken here. I'm aware of what the execute function does and that it is the standard way of working in Koheesio. Wat was throwing me off was the function called `func` on line 215. It is a strange name if you ask me. I initially thought it was a obscure internal function from the Koheesio library that I didn't know off. But then I see it defined higher up in the file. I just thought `download_files_in_partition` instead of `func` would be more descriptive and easier to understand.